File: android/gemcore/src/main/kotlin/com/gemwallet/android/domains/perpetual/PerpetualConfig.kt
package com.gemwallet.android.domains.perpetual

import uniffi.gemstone.Config

object PerpetualConfig {
    private val config get() = Config().getPerpetualConfig()

    val defaultLeverage: Int get() = config.defaultLeverage.toInt()

    val leverageOptions: List<Int> get() = config.leverageOptions.toUnsignedInts()

    val takeProfitOptions: List<Int> get() = config.takeProfitPercentOptions.toUnsignedInts()

    val stopLossOptions: List<Int> get() = config.stopLossPercentOptions.toUnsignedInts()

    val defaultTakeProfit: Int get() = config.defaultTakeProfitPercent.toInt()

    val defaultStopLoss: Int get() = config.defaultStopLossPercent.toInt()

    fun autocloseSuggestions(leverage: Int): List<Int> =
        Config().getAutocloseSuggestions(leverage.toUByte()).toUnsignedInts()

    fun selectLeverage(desired: Int, from: List<Int>): Int =
        Config().selectLeverage(
            desired.toUByte(),
            from.map { it.toByte() }.toByteArray(),
        ).toInt()
}

private fun ByteArray.toUnsignedInts(): List<Int> = map { it.toUByte().toInt() }

File: ios/Packages/GemstonePrimitives/Sources/PerpetualConfig.swift
// Copyright (c). Gem Wallet. All rights reserved.

import Foundation
import class Gemstone.Config

public struct PerpetualConfig {
    private init() {}

    public static var defaultLeverage: UInt8 {
        Config.shared.getPerpetualConfig().defaultLeverage
    }

    public static var leverageOptions: [UInt8] {
        Array(Config.shared.getPerpetualConfig().leverageOptions)
    }

    public static var takeProfitOptions: [UInt8] {
        Array(Config.shared.getPerpetualConfig().takeProfitPercentOptions)
    }

    public static var stopLossOptions: [UInt8] {
        Array(Config.shared.getPerpetualConfig().stopLossPercentOptions)
    }

    public static func selectLeverage(desired: UInt8, options: [UInt8]) -> UInt8 {
        Config.shared.selectLeverage(desired: desired, options: Data(options))
    }

    public static func autocloseSuggestions(leverage: UInt8) -> [UInt8] {
        Array(Config.shared.getAutocloseSuggestions(leverage: leverage))
    }
}
